src/cpp/session/modules/rmarkdown/NotebookAlternateEngines.cpp

```diff
@@ -452,63 +456,157 @@ Error runUserDefinedEngine(const std::string& docId,
+               nbCtxId,
+               ChunkOutputText);
+      
+      error = targetPath.parent().ensureDirectory();
+      if (error)
+         return error;
+
+      error = writeConsoleOutput(
+               kChunkConsoleOutput,
+               text,
+               targetPath,
+               true);
+      if (error)
+         return error;
+
+      enqueueChunkOutput(
+               docId,
+               chunkId,
+               nbCtxId,
+               ordinal++,
+               ChunkOutputText,
+               targetPath);
 
+      return Success();
+   };
+   
+   // TODO: do we need to capture console output? we already have a console
+   // output handler in the reticulate engine so we might want to avoid
+   // duplication of captured output, but should verify that errors and such
+   // are appropriately captured and reported
+   
    // run the user-defined engine
    SEXP outputSEXP = R_NilValue;
    Protect protect;
    error = RFunction(".rs.runUserDefinedEngine")
          .addParam(engine)
+         .addParam(code)
          .addParam(options)
          .call(&outputSEXP, &protect);
    
    // report errors during engine execution to user
    if (error)
    {
-      chunkConsoleOutputHandler(module_context::ConsoleOutputError,
-                                r::endUserErrorMessage(error),
-                                targetPath);
+      FilePath targetPath = module_context::tempFile(
+               "reticulate-engine-",
+               ".txt");
+      
+      chunkConsoleOutputHandler(
+               module_context::ConsoleOutputError,
+               r::endUserErrorMessage(error),
+               targetPath);
       return error;
    }
    
-   // forward engine output
+   // generic engine output (as a single string of console output)
    if (isString(outputSEXP))
    {
-      // write generated output to file
-      appendConsoleOutput(
-               kChunkConsoleOutput,
-               asString(outputSEXP),
-               targetPath);
-      
-      // default output is just text, for forward that
-      enqueueChunkOutput(
-               docId,
-               chunkId,
-               nbCtxId,
-               0,
-               ChunkOutputText,
-               targetPath);
+      emitText(asString(outputSEXP));
+   }
+   
+   // evaluate-style (list) output
+   else if (isList(outputSEXP))
+   {
+      int n = length(outputSEXP);
+      for (int i = 0; i < n; i++)
+      {
+         SEXP elSEXP = VECTOR_ELT(outputSEXP, i);
+         
+         // determine the output type based on the type of the current element
+         if (isString(elSEXP))
+         {
+            Error error = emitText(asString(elSEXP));
+            if (error)
+            {
+               LOG_ERROR(error);
+               continue;
+            }
+         }
+         else if (inherits(elSEXP, "condition"))
+         {
+            std::string message;
+            Error error = RFunction("base:::conditionMessage")
```

Comment:
When does this happen? Wondering if it can/should/does invoke the condition handling in notebooks (see https://github.com/rstudio/rstudio/blob/7317a5b2f036cc09a69fa182ad34ecd38ad13615/src/cpp/session/modules/rmarkdown/NotebookConditions.cpp#L34)

Reply:
The `reticulate` engine will itself capture R conditions, and return them as condition objects (or, more generally, anything using `evaluate::evaluate()`). This code unwraps the message associated with those condition objects so they can be reported back to the user.

We do this rather than installing a condition handler to ensure that output is fed back in the right order, as otherwise I think output / error / messages could get emitted out-of-order.

Reply:
Gotcha. Since conditions are being handled here, do we need (or want) to respect the error/warning chunk options as we do for conditions signaled from R code? For example:

https://github.com/rstudio/rstudio/blob/bc0cd9b715a181c52cd9c2759b1abbc9025187fd/src/cpp/session/modules/rmarkdown/NotebookExec.cpp#L240-L254

Reply:
Good idea -- I'll implement that!